src/modules/aws_ec2/mappers/instance_block_device_mapping.ts

```diff
@@ -0,0 +1,282 @@
+        if (e.region !== e.instance.region) continue;
+
+        // if instance is not created we are in the first step, no need to create anything
+        if (!e.instance?.instanceId) continue;
+        const client = (await ctx.getAwsClient(e.region)) as AWS;
+
+        // read volume details
+        if (!e.volumeId) throw new Error('Cannot attach empty volumes to an instance already created');
+        if (!e.volume?.volumeId) throw new Error('Tried to attach an unexisting volume');
+        if (e.volume.region !== e.instance.region)
+          throw new Error('Cannot create a mapping between different regions');
+
+        // only attach if volume is not in use. We can have the case that the volume is auto-attached on creation
+        if (e.volume.state === VolumeState.AVAILABLE) {
+          await this.attachVolume(client.ec2client, e.volume.volumeId, e.instance.instanceId, e.deviceName);
+        }
+
+        // return mapping
+        const newMap: InstanceBlockDeviceMapping = {
+          deviceName: e.deviceName,
+          instanceId: e.instanceId,
+          instance: e.instance,
+          volumeId: e.volumeId,
+          volume: e.volume,
+          region: e.instance.region,
+          deleteOnTermination: e.deleteOnTermination,
+        };
+
+        // Save the record back into the database to get the new fields updated
+        newMap.id = e.id;
+        await this.module.instanceBlockDeviceMapping.db.update(newMap, ctx);
+        out.push(newMap);
+      }
+      return out;
+    },
+    read: async (ctx: Context, id?: string) => {
+      if (id) {
+        // decompose the id
+        const { instanceId, volumeId, region } = this.idFields(id);
+        const client = (await ctx.getAwsClient(region)) as AWS;
+
+        // check if we can find the instance in database
+        const instance = await this.module.instance.db.read(
+          ctx,
+          this.module.instance.generateId({ instanceId: instanceId ?? '', region }),
+        );
+
+        if (instance) {
+          // read the instance mapping
+          const mapping = await this.module.instance.getInstanceBlockDeviceMapping(
+            client.ec2client,
+            instanceId,
+          );
+          for (const newMap of mapping ?? []) {
+            if (newMap.DeviceName && newMap.Ebs?.VolumeId === volumeId) {
+              const res = await this.instanceBlockDeviceMappingMapper(newMap, instance, ctx);
+              return res;
+            }
+          }
+        }
+      } else {
+        const out: InstanceBlockDeviceMapping[] = [];
+        const enabledRegions = (await ctx.getEnabledAwsRegions()) as string[];
+        await Promise.all(
+          enabledRegions.map(async region => {
+            const client = (await ctx.getAwsClient(region)) as AWS;
+            const rawInstances = (await this.module.instance.getInstances(client.ec2client)) ?? [];
+            for (const i of rawInstances) {
```

Comment:
I do not understand the reasoning behind using the `getInstances` SDK call directly here. It seems like you could just use `this.module.instance.db.read(ctx)` and skip the conditional `continue` calls before *and* skip the `await this.module.instance.db.read(ctx, ...generateId(...)` step below.